lib/auth/auth.go

```diff
@@ -654,7 +655,15 @@ func (a *Server) runPeriodicOperations() {
 		FirstDuration: utils.HalfJitter(time.Second * 10),
 		Jitter:        retryutils.NewHalfJitter(),
 	})
-	defer releaseCheck.Stop()
+	defer localReleaseCheck.Stop()
+
+	// reasonably small interval to ensure that users observe clusters as online within 1 minute of adding them.
+	remoteClustersRefresh := interval.New(interval.Config{
+		Duration: time.Second * 40,
```

Comment:
If I understand it correctly, we are changing an on-demand operation (GetRemoteCluster/GetRemoteClusters calls) to a periodic update. Do we have a grasp on what consequences could arise from this? (Not that I'm defending the current model, this seems a much saner behavior.)

For example, what would happen to a customer that runs 1000s of trusted clusters? (I do think we have at least one that does that.)

Also, why 40s?

Reply:
The jittered 40s was suggested by @rosstimothy in https://github.com/gravitational/teleport/issues/22008. The way I understand it, the goal is to make sure remote cluster statuses refresh under a minute.

Some consequences of updating 1000s of remote clusters periodically that I can think of:

### Bursts of backend reads/writes.

In my understanding, customers at this scale should already have their backend provisioned for high throughput. Also, the caching layer should help with reads.

### Schedule skew

Since all periodic tasks execute sequentially, updating many remote clusters can skew the schedule of other high-frequency tasks. I think it would help to run `refreshRemoteClusters()` in a goroutine with some precautions to avoid overlap if it takes more than 40s.

Reply:
Re Bursts: true, but emphasis on "should". It's never great when we break things that were working.

Re Schedule skew: yep, all you said there.

My concern here is that this change may not be as harmless as it looks. Worst case, we could be causing a periodic M*N "burst" of refreshes in scenarios with extreme numbers of trusted clusters. I'd be more amenable to taking the PR if we had benchmarks for those scenarios (which is a lot of work), or if we'd gone with a solution where chatter is minimal (an RFD starts to be desirable at this point).

Alternatively, if we can get approvals from both @rosstimothy and someone from the scalability team, then I'm happy to trust their judgment.

Reply:
Addressed the schedule skew concern in https://github.com/gravitational/teleport/pull/22455/commits/8e6c859ae6c65da65b622de86efa5ea0faf1bf84

Reply:
Skew changes LGTM.

Reply:
I think it'd be almost impossible for this to not be an improvement given how terrible updating on every read is, but having an upper-limit on the worst-case impact this has would be good.

I like the idea of limiting our peak writes.  Spikes in write-load is definitely a potential outage/degradation issue, and we might as well future-proof this while we're here.

A strategy we recently took with another periodic that I liked was to define a default write limit, but allow bumping that write limit when we have a massive amount of resources to update s.t. we still have a reasonable "worst case" time to all the resources being updated.  E.g. if we are ticking every 20s and want to ensure that worst-case we always update all resources within 10m, we could ensure that we always use a bucket size larger than 1/30th the total resource count.  Ex:

```golang
limit := 200
if dynamicLimit := (len(resources) / 30) + 1; dynamicLimit > limit {
    limit = dynamicLimit
}
// ...
```

I'm not sure we'd even need to do anything fancy like tracking where in the set of clusters we left off between ticks... we could just shuffle the slice on load, and then exit the function after `limit` number of writes.  Given the fact that most folks would have auth servers running in parallel, randomly selecting RCs in need of update might even be more optimal.

Reply:
Nice idea, thank you! I gave it a shot and implemented in https://github.com/gravitational/teleport/pull/22455/commits/46d6699036b38fa37ff5e42ccf1a762caf285cd1

So with the tick interval of 40s, default limit of 50 and bucket divider of 12 we have the worst case update delay of 8 minutes and the increased limit starts to kick in when the number of clusters is > 600.

Let me know if you think we should bump some of those.

Reply:
Thanks everyone for the interesting discussion and to Andrey for chasing this to the end. LGTM.